stdlib/public/core/Array.swift

```diff
@@ -1913,183 +1913,110 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
+      self[i].move(along: direction.elements[i])
     }
   }
 }
 
-/// Makes `Array` differentiable as the product manifold of `Element`
-/// multiplied with itself `count` times.
-extension Array : Differentiable where Element : Differentiable {
-  // In an ideal world, `TangentVector` would be `[Element.TangentVector]`.
-  // Unfortunately, we cannot conform `Array` to `AdditiveArithmetic` for
-  // `TangentVector` because `Array` already has a static `+` method with
-  // different semantics from `AdditiveArithmetic.+`. So we use
-  // `Array.DifferentiableView` for all these associated types.
-  public typealias TangentVector =
-    Array<Element.TangentVector>.DifferentiableView
-
-  public mutating func move(along direction: TangentVector) {
-    var view = DifferentiableView(self)
-    view.move(along: direction)
-    self = view.base
+extension Array.TangentVector: Equatable
+  where Array.Element.TangentVector: Equatable {
+  @inlinable
+  public static func == (lhs: Self, rhs: Self) -> Bool {
+    return lhs.elements == rhs.elements
   }
 }
 
-extension Array where Element : Differentiable {
-  public func _vjpSubscript(index: Int) ->
-    (Element, (Element.TangentVector) -> TangentVector)
-  {
-    func pullback(_ gradientIn: Element.TangentVector) -> TangentVector {
-      var gradientOut = Array<Element.TangentVector>(
-        repeating: .zero,
-        count: count)
-      gradientOut[index] = gradientIn
-      return TangentVector(gradientOut)
+extension Array where Element: Differentiable {
+  @inlinable
+  func _vjpSubscript(
+    index: Int
+  ) -> (Element, (Element.TangentVector) -> TangentVector) {
+    func pullback(_ v: Element.TangentVector) -> TangentVector {
+      var result = [Element.TangentVector](repeating: .zero, count: count)
+      result[index] = v
+      return TangentVector(result)
     }
     return (self[index], pullback)
   }
 
-  public static func _vjpPlus(_ lhs: [Element], _ rhs: [Element]) ->
-    ([Element], (TangentVector) -> (TangentVector, TangentVector)) {
-      func pullback(_ gradientIn: TangentVector) ->
-        (TangentVector, TangentVector) {
-        precondition(
-          gradientIn.base.count == lhs.count + rhs.count,
-          "+ should receive gradient with count equal to sum of operand " +
-            "counts, but counts are: gradient \(gradientIn.base.count), " +
-            "lhs \(lhs.count), rhs \(rhs.count)")
-        return (
-          TangentVector(Array<Element.TangentVector>(
-            gradientIn.base[0..<lhs.count])),
-          TangentVector(Array<Element.TangentVector>(
-            gradientIn.base[lhs.count...])))
-      }
-      return (lhs + rhs, pullback)
+  @inlinable
+  static func _vjpConcat(
+    _ lhs: [Element], _ rhs: [Element]
+  ) -> ([Element], (TangentVector) -> (TangentVector, TangentVector)) {
+    func pullback(_ v: TangentVector) -> (TangentVector, TangentVector) {
+      precondition(v.elements.count == lhs.count + rhs.count)
```

Comment:
Why did you not keep the precondition string `... should receive gradient with count equal to sum of...` in the precondition?